[nrfconnect] Added support for user data in Factory Data parser (#85)
ArekBalysNordic/connectedhomeip: docs/guides/nrfconnect_factory_data_configuration.md

-   [Enabling factory data support](#enabling-factory-data-support)
-   [Generating factory data](#generating-factory-data)
    -   [Creating factory data JSON file with the first script](#creating-factory-data-json-file-with-the-first-script)

>     -   [How to set user data](#how-to-set-user-data)

Reviewer: ```suggestion
    -   [How to set user data](#how-to-set-user-data)
        - [How to handle user data](#how-to-handle-user-data)
```

|  `discriminator`  |            Discriminator             |    2 B     |    uint16    |  mandatory  |                                                                                                                                                   A 12-bit value matching the field of the same name in the setup code. The discriminator is used during the discovery process.                                                                                                                                                    |
|    `passcode`     |            SPAKE passcode            |    4 B     |    uint32    |  optional   | A pairing passcode is a 27-bit unsigned integer which serves as a proof of possession during the commissioning. Its value must be restricted to the values from `0x0000001` to `0x5F5E0FE` (`00000001` to `99999998` in decimal), excluding the following invalid passcode values: `00000000`, `11111111`, `22222222`, `33333333`, `44444444`, `55555555`, `66666666`, `77777777`, `88888888`, `99999999`, `12345678`, `87654321`. |

> |      `user`       |              User data               |  variable  | JSON string  | max 1024 B  |                                                                          The user data is provided in the JSON format. This parameter is optional and depends on user's or manufacturer's purpose (or both). It is provided as a string from persistent storage and should be parsed in the user application. This data is not used by the Matter stack. To learn how to work with user data see [How to set user data](#how-to-set-user-data) section.                                                                          |

Reviewer 2: User here means manufacturer device manufacturer, right? If so, "user's or manufacturer's purpose (or both)" seems redundant. Also, is "It is provided as a string from persistent storage and should be parsed in the user application." up to date? In persistent storage it is stored as a CBOR map, not a string.

> location as an existing file. To allow overwriting, add the `--overwrite`
> option to the argument list of the Python script.

### How to set user data

> The user data is an optional field provided in factory data JSON file and depends on user's or manufacturer's purpose (or both). Currently, we support two data types that can be placed in user data field: `string` and `int32`.

Reviewer 2: We seem to lack the description that user is a JSON map whose values are restricted to strings and int32. Without that info a user might think that `user` itself should be string or int32.

To add user data as an argument to the [generate_nrfconnect_chip_factory_data.py](../../scripts/tools/nrfconnect/generate_nrfconnect_chip_factory_data.py) script add the following line to the argument list:

```
--user-data {user data JSON}
```

> As `user data JSON` you provide a single-level JSON file that consists of string or int32 fields. For example the following JSON file:

Reviewer 2: ```suggestion
As `user data JSON`,  provide a flat JSON map with values of `string` or `int32` type. ....
```

> location as an existing file. To allow overwriting, add the `--overwrite`
> option to the argument list of the Python script.

### How to set user data

The user data is an optional field provided in the factory data JSON file and depends on the manufacturer's purpose. The `user` field in a JSON factory data file is represented by a flat JSON map and it can consist of `string` or `int32` data types only. On the device side, the `user` data will be available as a CBOR map containing all defined `string` and `int32` fields.

To add user data as an argument to the [generate_nrfconnect_chip_factory_data.py](../../scripts/tools/nrfconnect/generate_nrfconnect_chip_factory_data.py) script, add the following line to the argument list:

```
--user-data {user data JSON}
```

> As `user data JSON`, provide a flat JSON map with values file that consists of `string` or `int32` types. For example, you can use a JSON file that looks like follows:

Reviewer: ```suggestion
As `user data JSON`, provide a flat JSON map with a value file that consists of `string` or `int32` types. For example, you can use a JSON file that looks like follows:
```

```
{
    "name": "product_name",
    "version": 123,
    "revision": "0x123"
}
```

When added to the argument line, the final result would look like follows:

```
--user-data {"name": "product_name", "version": 123, "revision": "0x123"}
```

> 
> The user data is not handled anywhere in the Matter stack, so you must handle it in your application. To do this, you can use the [Factory Data Provider](../../src/platform/nrfconnect/FactoryDataProvider.h) and apply one of the following methods:

Reviewer: ```suggestion

#### How to handle user data

The user data is not handled anywhere in the Matter stack, so you must handle it in your application. To do this, you can use the [Factory Data Provider](../../src/platform/nrfconnect/FactoryDataProvider.h) and apply one of the following methods:
```
